cpp/src/parquet/geometry_util_internal.h

```diff
@@ -0,0 +1,691 @@
+    }
+
+    ss << "]";
+
+    return ss.str();
+  }
+
+  Dimensions::dimensions dimensions;
+  double min[4];
+  double max[4];
+};
+
+inline bool operator==(const BoundingBox& lhs, const BoundingBox& rhs) {
+  return lhs.dimensions == rhs.dimensions &&
+         std::memcmp(lhs.min, rhs.min, sizeof(lhs.min)) == 0 &&
+         std::memcmp(lhs.max, rhs.max, sizeof(lhs.max)) == 0;
+}
+
+template <Dimensions::dimensions dims, bool swap, uint32_t chunk_size>
+class WKBSequenceBounder {
+ public:
+  explicit WKBSequenceBounder(double* chunk) : box_(dims), chunk_(chunk) {}
+  WKBSequenceBounder(const WKBSequenceBounder&) = default;
+
+  void ReadPoint(WKBBuffer* src) {
+    constexpr uint32_t coord_size = Dimensions::size<dims>();
+    src->ReadDoubles<swap>(coord_size, chunk_);
+    for (uint32_t dim = 0; dim < coord_size; dim++) {
+      if (ARROW_PREDICT_TRUE(!std::isnan(chunk_[dim]))) {
+        box_.min[dim] = std::min(box_.min[dim], chunk_[dim]);
+        box_.max[dim] = std::max(box_.max[dim], chunk_[dim]);
+      }
+    }
+  }
+
+  void ReadSequence(WKBBuffer* src) {
+    constexpr uint32_t coord_size = Dimensions::size<dims>();
+    constexpr uint32_t coords_per_chunk = chunk_size / coord_size;
+
+    uint32_t n_coords = src->ReadUInt32<swap>();
+    uint32_t n_chunks = n_coords / coords_per_chunk;
+    for (uint32_t i = 0; i < n_chunks; i++) {
+      src->ReadDoubles<swap>(coords_per_chunk * coord_size, chunk_);
+      ReadChunk(coords_per_chunk);
+    }
+
+    uint32_t remaining_coords = n_coords - (n_chunks * coords_per_chunk);
+    src->ReadDoubles<swap>(remaining_coords * coord_size, chunk_);
+    ReadChunk(remaining_coords);
+  }
+
+  void ReadRings(WKBBuffer* src) {
+    uint32_t n_rings = src->ReadUInt32<swap>();
+    for (uint32_t i = 0; i < n_rings; i++) {
+      ReadSequence(src);
+    }
+  }
+
+  void Reset() { box_.Reset(); }
+
+  void Finish(BoundingBox* out) { out->Merge(box_); }
+
+ private:
+  BoundingBox box_;
+  double* chunk_;
+
+  void ReadChunk(uint32_t n_coords) {
+    constexpr uint32_t coord_size = Dimensions::size<dims>();
+    for (uint32_t dim = 0; dim < coord_size; dim++) {
+      for (uint32_t i = 0; i < n_coords; i++) {
+        box_.min[dim] = std::min(box_.min[dim], chunk_[i * coord_size + dim]);
+        box_.max[dim] = std::max(box_.max[dim], chunk_[i * coord_size + dim]);
+      }
+    }
+  }
+};
+
+// We could (should?) avoid this madness by not templating the WKBSequenceBounder
+class WKBGenericSequenceBounder {
+ public:
+  WKBGenericSequenceBounder()
+      : chunk_{0.0},
+        xy_(chunk_),
+        xyz_(chunk_),
+        xym_(chunk_),
+        xyzm_(chunk_),
+        xy_swap_(chunk_),
+        xyz_swap_(chunk_),
+        xym_swap_(chunk_),
+        xyzm_swap_(chunk_) {}
+  WKBGenericSequenceBounder(const WKBGenericSequenceBounder&) = default;
+
+  void ReadPoint(WKBBuffer* src, Dimensions::dimensions dimensions, bool swap) {
+    if (ARROW_PREDICT_TRUE(!swap)) {
```

Comment:
are there benchmarks that show the swap template variable adds a lot of perfomance.  I'd expect branch prediction to do a pretty good job here?

Reply:
reading the comment above, I see that might be an open question.